apps/livreGUI/animationController/AnimationController.cpp

```diff
@@ -0,0 +1,275 @@
+ * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+ */
+
+#include <livre/core/types.h>
+
+#include <livreGUI/animationController/AnimationController.h>
+#include <livreGUI/Controller.h>
+
+#include <zeq/types.h>
+#include <zeq/event.h>
+#include <zeq/subscriber.h>
+#include <zeq/publisher.h>
+#include <zeq/hbp/vocabulary.h>
+
+namespace livre
+{
+
+const size_t DEFAULT_MIN_VALUE   = 0;
+const size_t DEFAULT_MAX_VALUE   = 1e4;
+const size_t DEFAULT_TICK_COUNTS = 10;
+
+/**
+ * AnimationEditor defines the possible actions for the animation feature
+ **/
+enum AnimationAction
+{
+    aa_pause,
+    aa_play,
+    aa_playBackwards
+};
+
+struct AnimationController::Impl
+{
+
+    Impl( AnimationController* animationController,
+          Controller& controller,
+          Ui_animationController& ui )
+        : _publisher( 0 )
+        , _subscriber( 0 )
+        , _ui( ui )
+        , _animationController( animationController )
+        , _controller( controller )
+        , _startFrame(DEFAULT_MIN_VALUE)
+        , _endFrame(DEFAULT_MAX_VALUE)
+        , _currentFrame(0)
+        , _action(aa_pause)
+        , _connected(false)
+        , _initialFrameReceived(false)
+    {
+    }
+
+    ~Impl()
+    {
+    }
+
+    void onFrame( const ::zeq::Event& event_ )
+    {
+        _initialFrameReceived = true;
+        const zeq::hbp::data::Frame frame
+                = zeq::hbp::deserializeFrame( event_ );
+        _startFrame = frame.start;
+        _endFrame  = frame.end;
+        _currentFrame = frame.current;
+        _ui.sldFrame->setMinimum( _startFrame );
+        _ui.sldFrame->setMaximum( _endFrame );
+
+        // Update slider value only when animation is active
+        if( _action == aa_play )
+            _ui.sldFrame->setValue( _currentFrame );
+    }
+
+    void setSubscriber()
+    {
+        const QString& uriStr = _ui.leURI->text();
+        const servus::URI uri( uriStr.toStdString( ));
+
+        _subscriber = _controller.getSubscriber( uri );
+        _subscriber->registerHandler( ::zeq::hbp::EVENT_FRAME,
+                                      boost::bind( &AnimationController::Impl::onFrame,
+                                                   this, _1 ));
+    }
+
+    void setPublisher()
+    {
+        const QString& uriStr = _ui.leURI->text();
+        const servus::URI uri( uriStr.toStdString( ));
+
+        _publisher = _controller.getPublisher( uri );
+    }
+
+    void connect()
+    {
+        try
+        {
+            setSubscriber();
+            setPublisher();
+
+            // Trying for 5 attempts to get a Frame event from the renderer. Only consider
+            // the connection as valid if the Frame event was received.
+            size_t attempts(5);
+            while( !_initialFrameReceived && attempts>0 )
+            {
+                _publisher->publish(
+                            ::zeq::vocabulary::serializeRequest(::zeq::hbp::EVENT_FRAME));
+                sleep(1);
```

Comment:
Hmm, so you have a subscriber, instead of busy waiting yo can connect when subscriber receives a message ?